Subscription.add should return the subscription.

It would be really nice if we could chain subscriptions with `.add()`.

Previously i wrote the following code:

```JS
function publish(source, selector) {
  return Rx.Observable.create(ob => {
    var xs = source.publish();
    return selector(xs).subscribe(ob).add(xs.connect());
  });
}
```

But this is actually wrong. It would be nice if this would be possible. Currently we need to write:

```JS
function publish(source, selector) {
  return Rx.Observable.create(ob => {
    var xs = source.publish();
    var sub = selector(xs).subscribe(ob)
    sub.add(xs.connect())
    return sub;
  });
}
```

**Update** 
We tried with the following:

```typescript
// addTo.ts
import {Subscription} from 'rxjs/Subscription';

export function addTo(collectorSub: Subscription): void {
  collectorSub.add(this);
}

export interface AddToSignature {
  (collectorSub: Subscription): void;
}
```

```typescript
// SubscriptionExtensions.ts
import {Subscription} from 'rxjs/Subscription';
import {addTo, AddToSignature} from './addTo';

declare module 'rxjs/Subscription' {
  interface Subscription {
    addTo: AddToSignature;
  }
}
```

which made VS Code happy, but then failed when building: 

> ERROR in XXXXXXX\node_modules\rxjs\Subject.d.ts
(9,22): error TS2420: Class 'Subject<T>' incorrectly implements interface 'Subscription'.
  Property 'addTo' is missing in type 'Subject<T>'.

--- a/spec/Subscription-spec.ts
+++ b/spec/Subscription-spec.ts
@@ -78,4 +78,83 @@ describe('Subscription', () => {
       done();
     });
   });
+
+  describe('Subscription.add()', () => {
+    it('Should returns the self if the self is passed', () => {
+      const sub = new Subscription();
+      const ret = sub.add(sub);
+
+      expect(ret).to.equal(sub);
+    });
+
+    it('Should returns Subscription.EMPTY if it is passed', () => {
+      const sub = new Subscription();
+      const ret = sub.add(Subscription.EMPTY);
+
+      expect(ret).to.equal(Subscription.EMPTY);
+    });
+
+    it('Should returns Subscription.EMPTY if it is called with `void` value', () => {
+      const sub = new Subscription();
+      const ret = sub.add(undefined);
+      expect(ret).to.equal(Subscription.EMPTY);
+    });
+
+    it('Should returns a new Subscription created with teardown function if it is passed a function', () => {
+      const sub = new Subscription();
+
+      let isCalled = false;
+      const ret = sub.add(function() {
+        isCalled = true;
+      });
+      ret.unsubscribe();
+
+      expect(isCalled).to.equal(true);
+    });
+
+    it('Should returns the passed one if passed a unsubscribed AnonymousSubscription', () => {
+      const sub = new Subscription();
+      const arg = {
+        isUnsubscribed: true,
+        unsubscribe: () => undefined,
+      };
+      const ret = sub.add(arg);
+
+      expect(ret).to.equal(arg);
+    });
+
+    it('Should returns the passed one if passed a AnonymousSubscription having not function `unsubscribe` member', () => {
+      const sub = new Subscription();
+      const arg = {
+        isUnsubscribed: false,
+        unsubscribe: undefined as any,
+      };
+      const ret = sub.add(arg as any);
+
+      expect(ret).to.equal(arg);
+    });
+
+    it('Should returns the passed one if the self has been unsubscribed', () => {
+      const main = new Subscription();
+      main.unsubscribe();
+
+      const child = new Subscription();
+      const ret = main.add(child);
+
+      expect(ret).to.equal(child);
+    });
+
+    it('Should unsubscribe the passed one if the self has been unsubscribed', () => {
+      const main = new Subscription();
+      main.unsubscribe();
+
+      let isCalled = false;
+      const child = new Subscription(() => {
+        isCalled = true;
+      });
+      main.add(child);
+
+      expect(isCalled).to.equal(true);
+    });
+  });
 });

--- a/src/Subscription.ts
+++ b/src/Subscription.ts
@@ -125,10 +125,12 @@ export class Subscription implements ISubscription {
    * list.
    */
   add(teardown: TeardownLogic): Subscription {
-    if (!teardown || (
-        teardown === this) || (
-        teardown === Subscription.EMPTY)) {
-      return;
+    if (!teardown || (teardown === Subscription.EMPTY)) {
+      return Subscription.EMPTY;
+    }
+
+    if (teardown === this) {
+      return this;
     }
 
     let sub = (<Subscription> teardown);

